### ISSUE TITLE
Files in dot-prefixed directories always Excluded.  Not in Ignore List.

### ISSUE THREAD

**Issue by a-raccoon** [hsoft/dupeguru#536]
I'm attempting to scan thousands of directories, and in them hundreds of directories that are prefixed by a dot (period) such as `.rsrc`.  dupeGuru insists on setting these directories to `Excluded` with no means for the user to disable this feature.  This pattern does NOT appear in the `Ignore List`.

Why? These directories were created by 7zip when extracting hundreds of PE .exe resources into directory structures.  The directory names are chosen by 7zip, not me.

**Comment by arsenetar** [comment]
You can change the state of these folders by finding them in the main window and changing them to normal.  It will include sub-directories of those directories in the search as long as they do not start with '.'.  These folders are normally private/hidden files which the user has not created so I imagine that might have been the original reason for setting these to Excluded by default.  It should be possible to create an option to include these directories as normal directories by default, or perhaps add the '.' pattern to the ignore path by default and let the user remove it instead of having this behavior hardcoded.

**Comment by arsenetar** [comment]
I am going to mark this as a feature as it is possible (although a bit inconvenient to search these folders), and this is going to require making either an option or a change in how this is setup which at this point feels more like a feature to me.

**Comment by a-raccoon** [comment]
I just figured out a solution in the current version.  Change all your added top-folders to Exclude, then change them back to Normal.  It will go through and set all your dot-folders from Exclude to Normal as well.

You can kill this bug without fixing.

**Comment by ghost** [comment]
> I just figured out a solution in the current version. Change all your added top-folders to Exclude, then change them back to Normal. It will go through and set all your dot-folders from Exclude to Normal as well.
> 
This doesn't work for me. Wonder what I am doing differently. I'm using the Linux version. I, too, would like an option where I can search all files regardless of whether or not they are hidden or system files. Maybe a "techy" mode or something.

**Comment by thinuspollard** [comment]
Is there any way we can have an option to include dotfiles and dot directories handled as normal files?

**Comment by glubsy** [comment]
I suppose we could remove [these lines](https://github.com/arsenetar/dupeguru/blob/4c3dfe2f1f11566534d3eb8d2d54bdcc688211ab/core/directories.py#L87-L90) and only rely on the Exclusion List (regular expressions).

**Comment by thinuspollard** [comment]
I think it does have a use case, in any case, people are expecting it to work this way.

Can we make it an option?

**Comment by arsenetar** [comment]
I am fine with removing the lines @glubsy pointed out or only using them if for some reason there is not an exclude list.  The exclude list by default adds those directories so the behavior by default would remain the same.  There is some state caching that is being done that does not always refresh correctly without additional changes though. If a folder has been loaded before they will "stick" to the non-Normal state once it has been loaded that way once.  So there is a bit more work that is needed to work out some oddities with that caching / reloading without cache when the exclude list is changed.

**Comment by DJCrashdummy** [comment]
> @a-raccoon Thanks for your solution. It would still be nice to have the option to change this behavior.

@Dobatymo it wouldn't "just be nice" to have an option to include/ignore dot folders and files but IMHO essential, because you can work around this issue for folders (really cumbersome if you have more than a few in the base folder), but **dot files in the base folder seem to be ignored in any case!** :open_mouth:
-> and this is IMHO a very bad practice, if not a bug: imagine a casual user who thinks **all files** were compared (because non were ignored neither in the list nor via options) and so he deletes the allegedly "empty" base folders manually. ...perhaps the hidden files were essential, have no duplicates and are now lost forever!

IMHO it doesn't have to be the default to include dot files/folders... *although personally i would prefer it this way: i rather see "too much" files at first and exclude them, instead of not seeing them and wondering what happened afterwards.*

**Comment by mosiser** [comment]
Does the "folder mode" consider dot files/folders? Or ignore them?

I have deleted some folders with "folder mode" and then, for safety, I extracted them from the trashbin and re-deleted with the "content mode".
As result, every dot-file wasn't deleted, the second time.
I don't know if the first time these folders was deleted (with contained dot-files) because the dot-files was checked and found identical, or because the scan ignore them...

**Comment by arsenetar** [comment]
@mosiser folder mode should compare the contents of the folders completely, this means all files within those folders will be considered.  Content mode will follow the exclusion filters and ignore lists when deciding which files to not scan.  By default the exclusion list contains a rule to ignore all files starting with `.` you can disable that rule if you would like.

**Comment by mosiser** [comment]
> folder mode should compare the contents of the folders completely, this means all files within those folders will be considered.

Thanks for clarification, I couldn't find this info anywhere...

> Content mode will follow the exclusion filters and ignore lists when deciding which files to not scan. By default the exclusion list contains a rule to ignore all files starting with `.` you can disable that rule if you would like.

Sorry, I didn't know how the Regex system works, I saw the exclusion list, but I didn't realise the dot-file exclusion was one of the options listed...
All right now. Thanks again.

**Comment by gitname** [comment]
> imagine a casual user who thinks all files were compared [...] and so he deletes the allegedly "empty" base folders manually

I did this. 😞  I did a Content scan under the assumption no files were being ignored. I didn't check the "Exclusion Filters" list (I had assumed it was empty by default).

It was only after I noticed some `.htaccess` files were _not_ included in the dupeGuru scan results--and then manually confirmed the files did exist in both scanned folders--that I searched for and found this Issue and learned about this default behavior.

One UI change that I think would make this default behavior more apparent to users is to have the "Exclusion Filters" tab be open by default (not necessarily _active_, but _open_ instead of closed) by default. I think that will make it more likely that new users check it.

![image](https://user-images.githubusercontent.com/7143133/182048879-1d7219f9-3a1b-4ad2-8804-2eda39182e44.png)

**Comment by gitname** [comment]
Here are the default "Exclusion Filters" entries, for reference:

![image](https://user-images.githubusercontent.com/7143133/182049401-52bb6b4b-d16f-4a3a-a3b6-d36d72c3583e.png)

The first one (`^\..*`) is the one that refers to dot-files.

Users may find it helpful to see a "Description" or "Comment" column (whose value is optional) in the table. For this particular entry, the "Description" value could be "Filename begins with dot".

**Comment by glubsy** [comment]
> have the "Exclusion Filters" tab be open by default (not necessarily active, but open instead of closed) by default

That's worth considering indeed.

The "exclusion filters" is a fairly advanced feature which was essentially implemented to circumvent the fact that dot files were excluded by default (it was hard coded if I remember correctly).

**Comment by jelabarre59** [comment]
I tried all the solutions (except for editing directories.py) above, and dupeguru is still excluding all the 'hidden' directories.

This is a problem, for example, if you're searching directories that have log files in the hidden dirs (such as Pidgin keeping it's log files in ~/.purple/logs)  If you're trying to clean up duplicates it won't work

**Comment by xho2** [comment]
> > imagine a casual user who thinks all files were compared [...] and so he deletes the allegedly "empty" base folders manually
> 
> I did this. 😞 I did a Content scan under the assumption no files were being ignored. I didn't check the "Exclusion Filters" list (I had assumed it was empty by default).

I just fell victim to the same. I often use `.`,`_`,`+` and other characters as prefixes for the purpose of alpha sorting or indicating a grouping type. Unfortunately, removing the filter still didn't resolve the issue for me, and as I couldn't find a way to fully expand the tree to double check I'd clicked every occurrence, I ended up going with a different software on this occasion..

As a side note, having just noticed this now, and knowing how long I've used these prefixes while also using DupeGuru...  
I feel a bit ill realizing how much I've likely inadvertently discarded.

**Comment by areddin** [comment]
I also encountered this issue. Clearing out an old hard drive and using DupeGuru to make sure that everything exists in another location before deleting. Worked great except that there were a bunch of directories that start with "." that contained real files. Disabling or removing the first item in the exclusions list ("^\..*") didn't change anything; the directories always showed up as "excluded" unless I manually changed each one (not practical in this case). Couldn't figure out another way, so I ended up batch renaming all the directories to remove the . to make it work.

**Comment by DJCrashdummy** [comment]
beforehand a short question @arsenetar:

> The exclude list by default adds those directories [starting with `.`] so the behavior by default would remain the same.

is it possible you made a slip of a pen here and meant files instead of directories? 🤔 
because how the help dialogue at the regex filters explains it, is that regex not containing a path separator will only get applied to files, hence the default `^\..*` should only consider files AFAIK.
...or perhaps something got lost in translation at the help dialogue?

**Comment by arsenetar** [comment]
@DJCrashdummy This change (which has never been put into a released build) https://github.com/arsenetar/dupeguru/commit/6d8b86b7eb1ae1763948b31dfba7a83dd5a8f1d7 is what "fixes" this behavior.  So using the code in the master branch would not have this issue given you use the new exclusion list feature.

**Comment by DJCrashdummy** [comment]
oh cool @arsenetar! sorry, i didn't notice that.

to be frank, i don't really know how the exclusion list works... i usually use the regex filter.

that's why i would have suggested beside removing the hardcoded filter to gain more flexibility (as the dot-files and -folders can still be covered by the regex filter), some additional measures...
1. change the default behavior - by also deactivating (not removing) the regex for hidden files - to completely rule out the issue of accidentally deleting hidden files/folders at least for new users.
2. add a hint to the help dialogue of the regex filters about clearing the cache might be necessary to apply the filters to all files/folders. ...as this might concern also other filters?!?
-> perhaps a different issue should be opened to address this bug in general. (...if it isn't yet?)
3. add a regex (but not activate it) to exclude dot-folders and - if no "explanation" column is added - mention at least these 2 for excluding dot-files/folders in the help dialogue.
so it shouldn't be hard also for people not knowing regex to still exclude all hidden items, **if** they really want to do so.

...i'm not 100% sure what still applies, as it seems that you already working on this/these issue(s). 🤷 - i just wanted to post my ideas to somehow fix this issue with somewhat low effort.

for completeness sake: i don't mind the other regex staying activated by default as they mostly concern cached files (perhaps even add some more for other cache files/folders would make sense)... because if they have duplicates, they mostly just spam your results and if you delete them by accident usually no harm is done. that can not be said for hidden files/folders in general! *(recycle bins are debatable... and depend IMHO on the users workflow. 😉)*
